services/spar/src/Spar/Sem/SAML2/Library.hs

```diff
@@ -0,0 +1,137 @@
+
+instance Member (Embed IO) r => HasNow (SPImpl r)
+
+instance Members '[Error SparError, Final IO, AReqIDStore] r => SPStoreID AuthnRequest (SPImpl r) where
+  storeID = (wrapMonadClientSPImpl .) . AReqIDStore.store
+  unStoreID = wrapMonadClientSPImpl . AReqIDStore.unStore
+  isAliveID = wrapMonadClientSPImpl . AReqIDStore.isAlive
+
+instance Members '[Error SparError, Final IO, AssIDStore] r => SPStoreID Assertion (SPImpl r) where
+  storeID = (wrapMonadClientSPImpl .) . AssIDStore.store
+  unStoreID = wrapMonadClientSPImpl . AssIDStore.unStore
+  isAliveID = wrapMonadClientSPImpl . AssIDStore.isAlive
+
+instance Members '[Error SparError, IdPEffect.IdP, Final IO] r => SPStoreIdP SparError (SPImpl r) where
+  type IdPConfigExtra (SPImpl r) = WireIdP
+  type IdPConfigSPId (SPImpl r) = TeamId
+
+  storeIdPConfig = SPImpl . App.runSparInSem . App.storeIdPConfig
+  getIdPConfig = SPImpl . App.runSparInSem . App.getIdPConfig
+  getIdPConfigByIssuerOptionalSPId a = SPImpl . App.runSparInSem . App.getIdPConfigByIssuerOptionalSPId a
+
+instance Member (Error SparError) r => MonadError SparError (SPImpl r) where
+  throwError = SPImpl . throw
+  catchError m handler = SPImpl $ catch (unSPImpl m) $ unSPImpl . handler
+
+saml2ToSaml2WebSso ::
+  forall r a.
+  Members
+    '[ AReqIDStore,
+       AssIDStore,
+       Error SparError,
+       IdPEffect.IdP,
+       Input Opts,
+       Logger String,
+       Embed IO,
+       Final IO
+     ]
+    r =>
+  Sem (SAML2 ': r) a ->
+  Sem r a
+saml2ToSaml2WebSso =
+  interpretH $ \case
+    AuthReq n ma i -> do
+      get_a <- runT ma
+      ins <- getInspectorT
+      x <- raise $ unSPImpl $ SAML.authreq @_ @SparError n (inspectOrBomb ins get_a) i
+      s <- getInitialStateT
+      pure $ x <$ s
+    AuthResp mitlt ma mb mc ab -> do
+      get_a <- runT ma
+      get_b <- runT mb
+      get_c <- bindT $ uncurry mc
+      ins <- getInspectorT
+      s <- getInitialStateT
+
+      x <- raise $ unSPImpl $ SAML.authresp mitlt (inspectOrBomb ins get_a) (inspectOrBomb ins get_b) (\x y -> inspectOrBomb ins $ get_c $ (x, y) <$ s) ab
+      pure $ x <$ s
+    Meta t ma mb -> do
+      get_a <- runT ma
+      get_b <- runT mb
+      ins <- getInspectorT
+      x <- raise $ unSPImpl $ SAML.meta t (inspectOrBomb ins get_a) (inspectOrBomb ins get_b)
+      s <- getInitialStateT
+      pure $ x <$ s
+    ToggleCookie sbs mp -> do
+      liftT $ unSPImpl $ SAML.toggleCookie sbs mp
+
+inspectOrBomb ::
+  Members
+    '[ AReqIDStore,
+       AssIDStore,
+       Error SparError,
+       IdPEffect.IdP,
+       Logger String,
+       Input Opts,
+       Embed IO,
+       Final IO
+     ]
+    r =>
+  Inspector f ->
+  Sem (SAML2 : r) (f b) ->
+  SPImpl r b
+inspectOrBomb ins get_a = do
+  fa <- SPImpl $ saml2ToSaml2WebSso get_a
+  maybe
+    (SPImpl . throw @SparError $ SAML.CustomError $ SparInternalError "saml2ToSaml2WebSso called with an uninspectable weaving functor")
```

Comment:
This can only possibly happen in the verdict handler, and I think it's fine.  Everything thrown via the error effects (mtl or polysemy) is caught before this, and everything thrown elsewhere is a legitimate internal error.